gems/smithy-client/lib/smithy-client/param_validator.rb

```diff
@@ -0,0 +1,190 @@
+# frozen_string_literal: true
+
+module Smithy
+  module Client
+    # @api private
+    class ParamValidator
+      include Smithy::Client::Shapes
+
+      EXPECTED_GOT = 'expected %s to be %s, got class %s instead.'
+
+      def self.validate!(rules, params)
+        new(rules).validate!(params)
+      end
+
+      def initialize(rules)
+        @rules = rules
+        # TODO: pump through options and consider this a flag?
+        @validate_required = false
+      end
+
+      # @param [Hash] params
+      # @return [void]
+      # @raise [ArgumentError] if the params are invalid
+      def validate!(params)
+        errors = []
+        structure(@rules, params, errors, 'params')
+        raise ArgumentError, error_messages(errors) unless errors.empty?
+      end
+
+      private
+
+      # TODO: depends on union implementation
+      # def validate_required_union(union_shape, values, errors)
+      #   set_values = values.to_h.length
+      #   if set_values > 1
+      #     errors << "multiple values provided to union at #{context} - must c
+      # ontain exactly one of the supported types: #{union_shape.member_names.join(', ')}"
+      #   elsif set_values == 0
+      #     errors << "No values provided to union at #{context} - must contain exactly
+      # one of the supported types: #{union_shape.member_names.join(', ')}"
+      #   end
+      # end
+
+      def structure(shape, values, errors, context)
+        case values
+        when Hash, shape.type then true
+        else
+          errors << expected_got(context, 'a hash', values)
+          return
+        end
+
+        validate_required_members(shape, values, errors, context) if @validate_required
+        # validate_required_union(shape, values, errors)
+
+        # validate non-nil members
+        values.each_pair do |name, value|
+          next if value.nil?
+
+          if shape.member?(name)
+            member_shape = shape.member(name)
+            shape(member_shape.shape, value, errors, context + "[#{name.inspect}]")
+          else
+            errors << "unexpected value at #{context}[#{name.inspect}]"
+          end
+        end
+      end
+
+      def list(shape, values, errors, context)
+        # ensure the value is an array
+        unless values.is_a?(Array)
+          errors << expected_got(context, 'an Array', values)
+          return
+        end
+
+        # validate members
+        values.each.with_index do |value, index|
+          next unless value
+
+          shape(shape.member.shape, value, errors, context + "[#{index}]")
+        end
+      end
+
+      def map(shape, values, errors, context)
+        unless values.is_a?(Hash)
```

Comment:
Ditto here - I think we could technically work with anything (ie a struct) that responds to `[]`

Reply:
I don't think we should allow a struct for maps. We don't even generate structs for maps. I think the majority of cases this would just be wrong and likely an issue with composing the request (maybe I forgot a key)